design.md

```diff
@@ -0,0 +1,106 @@
+
+`--remove` (shorthand `-r`) Remove a manifest from the index.
+
+**Note: One of the above flags should be used, as there has to be something to update. Otherwise the command does nothing.**
+
+`--oci-layout` Set the target as an oci image layout.
+
+`--tag` Tag the updated index. Multiple tags can be provided.
+
+(May be added in the future) `--output` Output the updated manifest to a location.
+
+#### Examples
+
+Add one manifest and remove two manifests from an index.
+
+`oras manifest index update localhost:5000/hello:latest --add win64 --remove sha256:xxx --remove arm64`
+
+## Issues and considerations
+1. **Command design: Making a subcommand group `index` under `oras manifest`(choose `oras manifest index create` instead of `oras manifest create-index` )`**
+
+Reasons:
+* The structure of the `oras manifest index` sub command group aligns well with the existing sub command groups `oras manifest/blob/repo`.
+* If in the future more index commands are needed, grouping them under the `index` group makes the manifest commands neater. We may also need operations for other manifest types, and creating new sub groups parallel to `index` looks feasible (i.e. `oras manifest image create`).
+
+2. **Combining "add manifest" and "remove manifest" operations as one `index update` operation**
+
+In my past iteration I created separate `oras index add-manifest` and `oras index remove-manifest` commands. After discussion with teammates, I think combining them as one `update` command is a better option, as it makes less garbage and fewer request calls when doing multiple adds and removes. But as a result, the command has many flags and the flags have complicated rules.
+
+3. **`oras manifest index create / update` will auto detect platform information for each source manifest. Currently we don't let user to specify platform unless we have a specific scenario in the future.**
+
+If we allow users to input platform information (introducing `--platform` for create and update commands), it will introduce a lot of complexity. Auto detecting platform information from config seems reasonable.
+
+4. **Use positional arguments to specify the target and sources, instead of using options**
+
+In the first iteration I used options to specify target index and its tag, making the list of manifests only command argument. Currently I use positional argument, so the first argument of the commands is the to-be-created/updated index with possible tags, and the remaining arguments are the list of manifests to add or remove. This aligns with many existing `oras` commands (`cp, attach, tag, manifest push`).
+
+5. **Should we require all the source manifests and the to-be-created index to be in the same repository?**
+
+Yes, as allowing multiple repositories will introduce a lot of copying of missing blobs and manifests.
+
+6. **Should we automatically push the created/updated index? Or should we decouple creating/updating the index and pushing the index?**
```

Comment:
(from Feynman) I would suggest "oras manifest create-index" only creates an image index in the local filesystem and let "oras manifest push" to push an index to the registry

Reply:
We may make the default behavior to be auto-push/not-push, and then provide a `--output` flag (in the future) to enable saving the index to the local filesystem.

Pro auto-push:
1. It aligns with the existing `oras attach` command, which automatically pushes the attached content.
2. For the case of oci layout, it's weird that we don't auto push the created index to the directory
3. It's easier to implement default auto push

Pro auto-not-push:
1. Users may want to modify the index (e.g. adding annotations) before pushing it to the repository.
2. Not auto push aligns with the behavior of tools like `podman`, `docker manifest create`